v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -329,73 +472,117 @@ async def connect(self):
+                        "The cancelled connect attempt may still complete as it is in-flight"
                     )
-                    self._mqtt_client.loop_stop()
-                    self._mqtt_client.loop_start()
-
-                # The result of the CONNACK is received via this future stored on the lock
-                rc = await self._connection_lock.future
-                # Sleep for 0.01 to briefly give up control of the event loop.
-                # This is necessary because a connect failure can potentially trigger both
-                # .on_connect() and .on_disconnect() and we want to allow them both to resolve
-                # before releasing the ConnectionLock.
-                await asyncio.sleep(0.01)
-                if rc != mqtt.CONNACK_ACCEPTED:
-                    logger.debug("Stopping Paho network loop due to connect failure")
-                    self._mqtt_client.loop_stop()
-                    raise MQTTConnectionFailedError(rc=rc)
+                    if reconnect_started_on_this_attempt:
+                        logger.debug(
+                            "Reconnect daemon was started with this connect attempt. Cancelling it."
+                        )
+                        self._reconnect_daemon.cancel()
+                        self._reconnect_daemon = None
+
+                        # NOTE: Because a connection could still complete after cancellation due to
+                        # it being in flight, this means that it's possible a connection could be
+                        # established without a running reconnect daemon, even if auto_reconnect
+                        # is enabled. This could be remedied fairly easily if so desired, but I've
+                        # chosen to leave it out for simplicity.
+                    else:
+                        logger.debug(
+                            "Reconnect daemon was started on a previous connect. Leaving it alone."
+                        )
+                    raise
+                finally:
+                    self._pending_connect = None
 
             else:
                 logger.debug("Already connected!")
 
+    async def _do_connect(self):
+        """Connect, start network loop, and wait for response"""
+
+        # Paho Connect
+        logger.debug("Attempting connect using port {}...".format(self._port))
+        try:
+            rc = await self._loop.run_in_executor(
+                None,
+                functools.partial(
+                    self._mqtt_client.connect,
+                    host=self._hostname,
+                    port=self._port,
+                    keepalive=self._keep_alive,
+                ),
+            )
+            rc_msg = mqtt.error_string(rc)
+            logger.debug("Connect returned rc {} - {}".format(rc, rc_msg))
+        # TODO: more specialization of errors to indicate which are/aren't retryable
```

Comment:
This can be done later - the infrastructure to support retryable vs nonretryable errors is implemented, I just haven't done an inventory of possible exceptions. I figure it will be quicker to do this while writing E2E tests rather than trying to do it now.